### DISCUSSION TITLE
Any interest in releasing core functions as a pypi module?

### DISCUSSION THREAD

**Discussion by songololo** [UMEP-dev/UMEP#662] [Ideas]
Thank you for a very interesting QGIS package.

Is there any possibility of providing the core functions as a pypi module for direct use with e.g. numpy etc? (i.e. skipping the QGIS ecosystem).

**Comment by biglimp** [comment]
Hi,
That would be a possibility. Some of the tools already exists. Is there any particular you are thinking of?

**Reply by songololo** [reply]
@biglimp it would be nice to be able to run the SOLWEIG workflow straight from Python on `numpy` arrays etc. So, things like the wall height and orientation, the SVF, and the SOLWEIG which uses these.

A cursory look through the code it seems possible to strip out some of the QGIS wrapping content and use the core functions directly. I've done a brief test on the wall functions and it seems doable, so would be nice if there were an official pypi package for accessing this without QGIS.

In our case we are interested in some automated Python workflows (educational context) to convert a geopackage to DSM, run the wall functions, SVF, SOLWEIG, then do some data takeoffs from the rasters for downstream comparison and analysis.

**Reply by biglimp** [reply]
I know that @matthiasdemuzere and @nilswallenberg was working on this. I dont know the status of this.

**Reply by songololo** [reply]
@biglimp we've managed to get an experimental / test version working with the code adapted for non-QGIS use. For now it includes:

- wall heights and aspects
- SVF
- Shadow processing
- SOLWEIG processing

Not all of the QGIS inputs are currently supported, for example, SOLWEIG workflow doesn't yet include landcover.

Are you interested in creating a new repository in `UMEP-dev` then we can move this code into the new repo and release as a pypi package (alpha or beta version for now)?

**Reply by biglimp** [reply]
Great! Can we have a look at the code? In a perfect world we should have a code (the core SOLWEIG code) that could be implemented in UMEP/QGIS. Than we only have one version to bugfix and update.

**Reply by songololo** [reply]
Sorry, only just had a chance to pick this up again, here is some example code:
https://github.com/songololo/umep-core

The README points to a demo file (the repo bundles some test data)

It would be nice to refactor some of the more intensive functions into a rust module.

**Reply by biglimp** [reply]
@songololo, sorry for this very late reply. I had a quick look at your repo and it looks very good. I would suggest that we create a new repo in `UMEP-dev` (as you suggested). That will also make it easier for us to implement updates and bugfixes in the models used. We could also add other models based on your code structure. We should also think on how to link it directly to the main code and functions in `UMEP for Processing` in a consistent way, so that we get rid of any unnecessary duplicates.

**Reply by songololo** [reply]
Thanks @biglimp adding you to the repo so that we can transfer ownership.

It would be nice to identify core functions and write modular tests for expected behaviour so that we can experiment with `rust` conversions.

**Reply by biglimp** [reply]
Accepted your invite. What would be the easiest way to transfer into `UMEP-dev` (never really learned how to use GH properly)? 

@nilswallenberg and @matthiasdemuzere, lets continue within `umep-core`, on the work of setting up stand-alone capabilities for parts of umep (SEBE, SpatialTC etc.). When it has been transferred, lets talk and see whats need to be done (e.g. land cover scheme etc.).

@j3r3m1 , could we also potentially include URock? Would be nice to have a full workflow all the way to UTCI, PET and COMFA.

**Reply by songololo** [reply]
@biglimp looking into transferring - are you able to add me to the UMEP-dev Github organisation? I can then transfer it directly. Alternatively, now that you've been added as a collaborator to the UMEP core repo you might have permissions to be able to transfer it directly from your account?

**Reply by biglimp** [reply]
@songololo , I think i messed up. Could you please invite me again for the ownership of `umep-core`.

**Reply by songololo** [reply]
@biglimp new invite sent!

**Reply by j3r3m1** [reply]
> @j3r3m1 , could we also potentially include URock? Would be nice to have a full workflow all the way to UTCI, PET and COMFA.

Sure, most of the URock code can already been called from outside QGIS so probably minor changes are needed. Lets discuss that when you will come to France.

**Reply by biglimp** [reply]
Ok, easier if I add you @songololo  to the UMEP-dev organisation. An invite will arrive soon...

**Reply by songololo** [reply]
@biglimp thanks - was able to transfer: https://github.com/UMEP-dev/umep-core

**Reply by biglimp** [reply]
Ok, lets move over to the new repository. I have started. I am working on a solweig workflow using the GBG datasets. Will upload when finished. Create issues as you go along.
